lib/docker-stdio.js

```diff
@@ -226,29 +233,65 @@ function createDockerFileReadStreamServer(opts, callback) {
         }
     }
 
-    /**
-     * Create TCP Server which will output the archive stream
-     */
-    tcpServer = net.createServer();
+    var loadOpts = {
+        log: opts.log,
+        req_id: opts.req_id,
+        uuid: opts.uuid
+    };
 
-    // Close server if no connections are received within timeout
-    var serverTimeout = setTimeout(function () {
-        log.warn('Closing stream tcpServer after ' +
-             timeoutSeconds + ' seconds without connection');
-        tcpServer.close();
-    }, timeoutSeconds * 1000);
+    async.waterfall([
+        /**
+         * Capture zone state, as this will dictate which method we use to do
+         * the copying.
+         */
+        function (next) {
+            vmadm.load(loadOpts, function (err, vm) {
+                if (err) {
+                    next(err);
+                    return;
+                }
+                zoneState = vm.zone_state;
+                next();
+            });
+        },
+
+        /**
+         * Create TCP Server which will output the archive stream
+         */
+        function (next) {
+            tcpServer = net.createServer();
+
+            // Close server if no connections are received within timeout
+            serverTimeout = setTimeout(function () {
+                log.warn('Closing stream tcpServer after ' +
+                     timeoutSeconds + ' seconds without connection');
+                tcpServer.close();
+            }, timeoutSeconds * 1000);
+
+            if (zoneState === 'running') {
+                tcpServer.on('connection', onConnectionZoneRunning);
+            } else {
+                tcpServer.on('connection', onConnectionZoneNotRunning);
+            }
+
+            tcpServer.on('error', function (err) {
+                log.error({ err: err }, 'read stream error');
```

Comment:
@notmatt @orlandov Yes, e.g if the server socket can't be created because the port is already in use, then `'listening'` will not be emitted, so the `'error'` event listener should also call `next`. Guarding against calling `next` twice might be needed too, I'm not sure what `async` does in that case.